### PR TITLE
linux-capture: Disable strict binding for NVIDIA drivers

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
NVIDIA drivers appears to have a bug where binding would be excessively
slow. Apply a workaround similar to what [KWin] does to prevent the issue.

Also performs a refactor so that the code paths with and without the
workaround can be shared.

[KWin]: https://github.com/KDE/kwin/blob/4f2c3a00c4a86c08321e80d601fecd09d0871c79/src/libkwineffects/kwinglplatform.cpp

Fixes: 316f858c6 ("linux-capture: Fix capturing on software rasterization setups")
Closes: https://github.com/obsproject/obs-studio/issues/5685

Tested-By: univrsal <uni@vrsal.xyz>

### Types of changes
Bug fix (non-breaking change which fixes an issue)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- N/A I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by ishitatsuyuki** [comment]
cc @kkartaltepe since you reviewed the previous PR

**Comment by kkartaltepe** [comment]
I think this looks good to me, but I wont be able to test it for another two weeks or so.

**Comment by kkartaltepe** [comment]
I still cant test this but I had a chance to look into some of the kwin code. It seems that most drivers should be using loose binding (at least kwin sets this for nvidia/amd and intel seems to need research). Since this ends up being a potential for issues we should probably modify the code to only use strict binding in the case of a software renderer and otherwise stick with our previous behavior of loose binding.

**Comment by ishitatsuyuki** [comment]
> It seems that most drivers should be using loose binding (at least kwin sets this for nvidia/amd and intel seems to need research).

If you are talking about the code I linked in the commit message, no, it doesn't set it for modern AMD cards. I just checked my logs and I get `Requires strict binding:                yes`.

It has caused trouble with Intel in the past (https://bugs.freedesktop.org/show_bug.cgi?id=80349), which is why there's a line setting it to false for Intel cards (the default is false anyway, the assignment could have been removed in the patch).

Since the default is false, it's using strict binding for any other vendors (qcom etc, they actually use GLES/EGL anyway). The remaining question is legacy AMD drivers, for which two variants exists:
- Proprietary drivers have been mostly retired due to binary incompatibility: they don't run with the latest Xorg.
- Non-gallium drivers have got [booted](https://www.phoronix.com/scan.php?page=news_item&px=Mesa-Classic-Retired) from Mesa. Nobody should be using them at this point.
- I tend to believe the Gallium drivers are OK these days, although since r600 is explicitly marked to use loose binding in KWin I can port that 10-years old check to this patch too.

**Comment by kkartaltepe** [comment]
> If you are talking about the code I linked in the commit message, no, it doesn't set it for modern AMD cards. I just checked my logs and I get Requires strict binding: yes.

Ah I see I guess r600 driver is the previous generation, since it looked like it was being set for those then AMD is split. However since there are no issues reported on AMD it seems that strict binding is not required so again my preference is to retain the old functionality unless there are known issues we can point to.

> It has caused trouble with Intel in the past (https://bugs.freedesktop.org/show_bug.cgi?id=80349), which is why there's a line setting it to false for Intel cards (the default is false anyway, the assignment could have been removed in the patch).

Right this patch is incredibly old so its unclear if this was a performance regression for the sake of this issue or not. Also we don't have this issue reported as far as I know so that is why my hesitant to say we should change our binding strategy on intel without more information.

> Since the default is false, it's using strict binding for any other vendors (qcom etc, they actually use GLES/EGL anyway).

This case (mesa drivers for all EGL users) ends up using loose binding it seems https://github.com/KDE/kwin/blob/4f2c3a00c4a86c08321e80d601fecd09d0871c79/src/libkwineffects/kwinglplatform.cpp#L1092 unless you mean proprietary drivers.

**Comment by ishitatsuyuki** [comment]
> However since there are no issues reported on AMD it seems that strict binding is not required so again my preference is to retain the old functionality unless there are known issues we can point to.

OK, I will add r600 to the loose binding special casing list.

> Right this patch is incredibly old so its unclear if this was a performance regression for the sake of this issue or not. Also we don't have this issue reported as far as I know so that is why my hesitant to say we should change our binding strategy on intel without more information.

We should change the strategy because the old one is not spec compliant. At least some of the Mesa developers are aware of this "correct" interpretation: https://oftc.irclog.whitequark.org/dri-devel/2021-11-07#1636283343-1636283065; so it's reasonable to expect that strict binding is not slow on Mesa drivers. FWIW on my AMD setup it had nearly zero overhead (XLock took more time than the binding operations).

> This case (mesa drivers for all EGL users) ends up using loose binding it seems KDE/kwin@4f2c3a0/src/libkwineffects/kwinglplatform.cpp#L1092 unless you mean proprietary drivers.

KWin is actually wrong in this case, the `OES_EGL_image_external` (which supersedes `OES_EGL_image`) spec says

        Sampling an external texture which has been modified since it was
        bound will return samples which may correspond to image values either
        before, during, or after the modification.  Binding (or re-binding if
        already bound) an external texture by calling BindTexture after all
        modifications are complete guarantees that sampling done in future
        draw calls will return values corresponding to the values in the
        buffer at or after the time that BindTexture is called.  (Note that
        calling BindTexture after calling commands that modify the EGLImage
        may not be sufficient to ensure that the modifications are complete;
        additional synchronization (for example eglWaitAPI, eglWaitNative,
        glFinish, vgFinish, or eglWaitSyncKHR, etc) may be required following
        the call initiating modifications, to ensure they have taken effect on
        the texture, before the call to BindTexture is made.)

Basically the same requirement as GLX. The demo package also got updated https://github.com/freedesktop/mesa-demos/commit/f84d2225208b9ccb0f4f027d72fa75bb25439920#diff-ca608d789dbf11c506f2d27011700674e68def4db4b82bc2af047fae3a5e85a0 at around the time, although the KWin patch precedes it but I can imagine it not being communicated due to a delayed release.

According to the KWin commit it's nouveau causing the headache in this case (https://github.com/KDE/kwin/commit/9e2b6edf02b47dc414dac4ce0dbb48c2d1b949e7), we don't support EGL right now but we'll probably need to dig that grave to decide on a strict binding policy when we deal with it...

**Comment by Techjar** [comment]
I was having serious performance issues with Xcomposite window capture on the latest OBS master. NVIDIA card here, of course. This PR brought it right back to nominal performance.

**Comment by LGCW** [comment]
Tested this on the box is the studio this morning after encountering the bug in the latest beta. Nvidia 2060 running 495.46, xfce4.

This PR fixes the issue with Xcomposite bringing everything to a crawl.

**Review by WizardCM — CHANGES_REQUESTED** [review]
This introduces no regressions on Intel, and fixes the performance regression on Nvidia.

However, it does seem to introduce a memory leak, counted on shutdown. This can be reproduced by having 1 window capture source and closing OBS, logging 2 memory leaks.

**Comment by ishitatsuyuki** [comment]
> However, it does seem to introduce a memory leak, counted on shutdown.

I cannot reproduce this with both radeonsi and llvmpipe (`LIBGL_ALWAYS_SOFTWARE=1`). Can you paste your log? I guess it's probably due to a bad error handling path, and I've seen instances of `glXReleaseTexImageEXT` failing on NV for unknown reasons.

**Comment by jp9000** [comment]
I'm not sure how this would add a memory leak. I don't see any allocations in this code.

**Comment by WizardCM** [comment]
Log with this PR:
[2022-01-02 12-24-58.txt](https://github.com/obsproject/obs-studio/files/7807016/2022-01-02.12-24-58.txt)

Log without this PR (master at the time of test):
[2022-01-02 12-28-32.txt](https://github.com/obsproject/obs-studio/files/7807018/2022-01-02.12-28-32.txt)

I reproduced it a second time relaunching without removing/readding the source (after switching back to this branch):
[2022-01-02 12-31-54.txt](https://github.com/obsproject/obs-studio/files/7807025/2022-01-02.12-31-54.txt)

It's entirely possible the leaks are unrelated & a side effect of the branch being based on an older `master`.

**Comment by ishitatsuyuki** [comment]
I honestly have no idea why there's a leak, I just rebased the branch so @WizardCM you may want to test again.

If it persists, maybe it's good time to compile with ASan to get allocations tracked with stack traces.

**Comment by kkartaltepe** [comment]
Tested the latest PR, i dont have any memory leaks on intel or nvidia. Performance seems fine on both.

**Review by kkartaltepe — APPROVED** [review]

**Comment by jp9000** [comment]
Nice branch name -- I sympathize with you as a Linux user. NVIDIA has not given me pleasant experiences on Linux.
